File: sdk/examples/client/stronghold.rs
// Copyright 2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we will create an address with a stronghold secret manager.
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --all-features --example stronghold
//! ```

use iota_sdk::{
    client::{
        constants::{IOTA_COIN_TYPE, IOTA_TESTNET_BECH32_HRP},
        secret::{stronghold::StrongholdSecretManager, PublicKeyOptions, SecretManageExt},
        Result,
    },
    crypto::keys::bip39::Mnemonic,
    types::block::address::{Ed25519Address, ToBech32Ext},
};

#[tokio::main]
async fn main() -> Result<()> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    #[allow(clippy::single_element_loop)]
    for var in ["MNEMONIC"] {
        std::env::var(var).unwrap_or_else(|_| panic!(".env variable '{var}' is undefined, see .env.example"));
    }

    let secret_manager = StrongholdSecretManager::builder()
        .password("some_hopefully_secure_password".to_owned())
        .build("test.stronghold")?;

    let mnemonic = Mnemonic::from(std::env::var("MNEMONIC").unwrap());

    // The mnemonic only needs to be stored the first time
    secret_manager.store_mnemonic(mnemonic).await?;

    // Generate address
    let address = secret_manager
        .generate::<Ed25519Address>(&PublicKeyOptions::new(IOTA_COIN_TYPE))
        .await?
        .to_bech32(IOTA_TESTNET_BECH32_HRP);

    println!("First public address: {address}");

    Ok(())
}

File: bindings/python/examples/secret_manager/stronghold.py
import os

from dotenv import load_dotenv

from iota_sdk import SecretManager, StrongholdSecretManager

load_dotenv()

for env_var in ['MNEMONIC', 'STRONGHOLD_PASSWORD']:
    if env_var not in os.environ:
        raise Exception(f'.env {env_var} is undefined, see .env.example')


secret_manager = SecretManager(StrongholdSecretManager(
    "example.stronghold", os.environ['STRONGHOLD_PASSWORD']))

# Store the mnemonic in the Stronghold snapshot, this needs to be done only the first time.
# The mnemonic can't be retrieved from the Stronghold file, so make a
# backup in a secure place!
secret_manager.store_mnemonic(
    os.environ['MNEMONIC'])

# Generate public address with custom account index and range.
address = secret_manager.generate_ed25519_addresses(
    account_index=0, start=0, end=1)

print(f'Address: {address[0]}')
